metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,569 @@
+      nameQualsList: List[ForYieldEnumeration],
+      forYieldConditionsList: List[ForYieldCondition],
+      yieldExpression: YieldExpression,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+
+    val enumerations = nameQualsList.flatMap { nameQual =>
+      for {
+        valName <- nameQual.perhapsVariableName
+        assignOrMap <- nameQual.perhapsAssignOrMap
+        qual <- nameQual.qual
+      } yield (valName, assignOrMap, qual)
+    }
+
+    val conditions = forYieldConditionsList.flatMap { condition =>
+      for {
+        conditionString <- condition.condition
+        filterOrNot <- condition.maybeFilterOrNot
+      } yield (filterOrNot, conditionString)
+
+    }
+
+    val forYieldString =
+      s"""|for {
+          |${enumerations.map(nameQual => s"${indentation}  ${nameQual._1} ${nameQual._2} ${nameQual._3}").mkString("\n")}
+          |${conditions.map(forYieldCondition => s"${indentation}  if ${forYieldCondition._1}(${forYieldCondition._2})").mkString("\n")}
+          |${indentation}} yield {
+          |${indentation}  ${yieldExpression.expression}
+          |${indentation}}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def maybeBuildToForYieldChainedCodeActionWithApply(
+      path: AbsolutePath,
+      termApply: Term.Apply,
+      document: String,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    val (nameQualsList, conditions, _) = {
+      extractChainedForYield(
+        None,
+        List.empty,
+        List.empty,
+        termApply,
+        document,
+        Set.empty
+      )
+    }
+
+    if (nameQualsList.nonEmpty) {
+      pprint.log("nameQualsList is: " + nameQualsList)
+      nameQualsList match {
+        case heads :+ tail =>
+          tail.qual.map(yieldQual =>
+            buildCodeActionFromForYieldParts(
+              heads,
+              conditions,
+              YieldExpression(yieldQual),
+              indentation,
+              path,
+              termApply.pos.toLSP.getStart,
+              termApply.pos.toLSP.getEnd
+            )
+          )
+        case _ :: Nil => None
+      }
+
+    } else None
+
+  }
+
+  private def maybeBuildToForYieldCodeActionWithSelect(
+      path: AbsolutePath,
+      termSelect: Term.Select,
+      document: String,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    termSelect.parent.collect { case termApply: Term.Apply =>
```

Comment:
So you want the for comprehension to happen on the uppermost parent in the chain?

Reply:
I think that's the only sensible solution, otherwise how does it currently work if you run the action on the last map or flatMap ?

Reply:
At first glance, this suggestion seems correct, but during the implementation I realized that the uppermost parent is actually the last map/flatMap in t, from which we start. This would also introduce the problem that if the last method is not map/flatMap/filter, no code action would be shown, which would be a bug.

So for example if the user has `list.map(…..).flatMap(…).filter(…).sort(….)` no code action would be shown due to `sort` because the upper most parent in the tree hierarchy is `sort`

Reply:
At first glance, this suggestion seems correct, but during the implementation I realized that the uppermost parent is actually the last map/flatMap in t, from which we start. This would also introduce the problem that if the last method is not map/flatMap/filter, no code action would be shown, which would be a bug.

So for example if the user has `list.map(…..).flatMap(…).filter(…).sort(….)` no code action would be shown due to `sort` because the upper most parent in the tree hierarchy is `sort`

Reply:
Let's start with converting only chains that fully can be changed to for comprehension. That's the base scenario that we should make work. 

Later we could also traverse up the parent,but only until the methods are ok.

Reply:
Currently, it is working for both of the cases you mentioned ;)